### PR TITLE
Fix nans in gradient from inverse softplus

### LINKED ISSUE #119: Clamping introduces `nan` if `ar_steps_train > 1`
After a few iterations clamping state variables introduces nan in the train loss. Replacing the sigmoid and inverse sigmoid functions with simple `torch.clamp` does prevent the issue (just as an indication). Here is a reproducible example based on the danra datastore from the `test_examples`: 
- First build the dataset and the graph (I built `hierarchical` archetype)
- Then start the training with neural_lam
``` bash
             python -m neural_lam.train_model \
            --config_path ./neural-lam/tests/datastore_examples/mdp/danra_100m_winds/config.yaml \
            --model hi_lam \
            --graph_name hierarchical \
            --hidden_dim 16 \
            --hidden_dim_grid 16 \
            --time_delta_enc_dim 16 \
            --processor_layers 2 \
            --batch_size 1 \
            --epochs 1 \
            --ar_steps_train 2 \
            --ar_steps_eval 2 \
            --val_steps_to_log 1 2
```
- Observe the introduction of nans
``` bash
Training: |          | 0/? [00:00<?, ?it/s]
Training:   0%|          | 0/20 [00:00<?, ?it/s]
Epoch 0:   0%|          | 0/20 [00:00<?, ?it/s] 
Epoch 0:   5%|▌         | 1/20 [00:02<00:49,  0.39it/s]
Epoch 0:   5%|▌         | 1/20 [00:02<00:49,  0.39it/s, v_num=pbts, train_loss_step=4.740]
Epoch 0:  10%|█         | 2/20 [00:03<00:27,  0.65it/s, v_num=pbts, train_loss_step=4.740]
Epoch 0:  10%|█         | 2/20 [00:03<00:27,  0.65it/s, v_num=pbts, train_loss_step=nan.0]
Epoch 0:  15%|█▌        | 3/20 [00:03<00:20,  0.81it/s, v_num=pbts, train_loss_step=nan.0]
Epoch 0:  15%|█▌        | 3/20 [00:03<00:21,  0.80it/s, v_num=pbts, train_loss_step=nan.0]
```

### PR DESCRIPTION
**Pull request by SimonKamuk**
## Describe your changes

Fix a bug where the inverse softplus causes gradients to become nans.

I believe the issue was that even though the `non_linear_part` had clamping inside, and the values were later being discarded if they were above the threshold, the input x values could still cause numerical instabilities before we checked if the input was above the threshold.

What I did to fix it was to clamp the input x values. The limits were previously defined as $exp(x\cdot beta)-1 \ge 10^{-6}$ and $x\cdot beta \le threshold$, so I changed it to the equivalent $log(10^{-6}+1)/beta \le x \le threshold/beta$.

## Issue Link

closes #119

## Type of change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [ ] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [ ] I have performed a self-review of my code
- [ ] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by SimonKamuk** [comment]
Since our tests now use the reduced danra domain and are relatively quick, what do you think about adding `detect_anomaly=True` to the `pl.Trainer` in `test_training.py` ? That would prevent an issue like this from getting through (assuming the training config used for the test includes any such future features). Is it a bit overkill?

**Review by sadamov — APPROVED** [review]
Tested this with two different datasets and ar_steps_train, solves the bug. Thanks Simon

**Comment by joeloskarsson** [comment]
I think it would be a good idea to add detect_anomaly=True to the tests
